proposals/lip-0052.md

```diff
@@ -679,6 +671,172 @@ else: # chainID == sendingChainID
+        }
+    }
+}
+```
+
+#### Genesis State Initialization
+
+During the genesis state initialization stage, the following steps are executed. If any step fails, the block is discarded and has no further effect.
+
+Let `genesisBlockAssetBytes` be the `data` bytes included in the block assets for the NFT module and let `genesisBlockAssetObject` be the deserialization of `genesisBlockAssetBytes` according to the `genesisNFTStoreSchema` schema, given above.
+
+* Initial checks:
+   * Across all elements of the `NFTSubstore` array, all values given for `NTFID` must be unique. 
+   * For all elements of the `NFTSubstore` array, values given for `chainID` must have either length 20 (representing a user address) or 4 (representing a chain ID).
+   * For all elements of the `NFTSubstore` array, if the value of `owner` has length 4 (representing an escrowed NFT) then the corresponding `NFTID.chainID` must equal `CHAIN_ID_ALIAS_NATIVE`.
+   * For all elements of the `NFTSubstore` array, if the value of `attributes` must have length less than or equal to `MAX_BYTE_LENGTH_ATTRIBUTES`.
+   * Accross all element of the `collectionSubstore` array, all values given for `collection` must be unique.
+
+* For each entry `NFTEntry` in `genesisBlockAssetObject.NFTSubstore`, create an entry in the NFT substore  with
+  ```python
+  storeKey = uint32be(NFTEntry.NFTID.chainID) || uint32be(NFTEntry.NFTID.collection) || uint64be(NFTEntry.NFTID.index)
+  storeValue = {"owner": NFTEntry.owner,
+                "lockingModuleID": NFTEntry.lockingModuleID,
+                "attributes": NFTEntry.attributes
+               } serialized using NFTStoreSchema.
+  ```
+  Further, if `NFTEntry.owner` has length 20, create an entry in the user substore with
+  ```python
+  storeKey = NFTEntry.owner || uint32be(NFTEntry.NFTID.chainID) || uint32be(NFTEntry.NFTID.collection) || uint64be(NFTEntry.NFTID.index)
+  storeValue = {"exists": True} serialized using userStoreSchema.
+  ```
+
+* For each entry `collectionEntry` in `genesisBlockAssetObject.collectionSubstore`, create an entry in the collection substore  with
+  ```python
+  storeKey = uint32be(collectionEntry.collection)
+  storeValue = {"nextAvailableIndex": collectionEntry.nextAvailableIndex} serialized using collectionStoreSchema.
+  ```
+
+* Create an entry in the available collection substore with
+  ```python
+  storeKey = EMPTY_BYTES
+  storeValue = {"nextAvailableCollection": genesisBlockAssetObject.availableCollectionSubstore} 
+               serialized using availableCollectionStoreSchema.
+  ```
+
+* For each entry `terminatedChainID` in `genesisBlockAssetObject.terminatedEscrowSubstore`, create an entry in the terminated escrow substore with
+  ```python
+  storeKey = uint32be(terminatedChainID)
+  storeValue = {"escrowTerminated": True} serialized using terminatedEscrowStoreSchema.
+  ```
+  
+Once the module store is initialized, its validity is attested asserting that the function below returns `True`.
```

Comment:
I was wondering, shouldn't this function execution be better placed in the *Genesis state finalization* processing step of the genesis block? This way we make sure this step is  always done after the others without creating inconsistencies. Also, we  would have a clear distinction between general checks and state initialization versos inter substore (although intramodule) validation. T

Reply:
No very strong opinion on that, I understood that we do all "internal" checks during the initialization step and keep the finalization step for cross-module API calls (or verifications).
Also the sentence above the code snippet states that this state validation must be done after the module store is created.

Reply:
I think that idea of intra vs inter module checks can be flexible and in [the genesis block LIP](https://github.com/LiskHQ/lips/blob/38458a4c0ac3c24ab517ad6ffb069b102137fc44/proposals/lip-update_genesis_block_schema_and_processing.md) this definition is relatively blurry on purpose. And this NFT module with its peculiar state store complexity may be suitable for such genesis processing.

Again, it is just a small consideration and as you said, the current specs are correct as they are. Also @janhack or @mitsuaki-u  may have better or stronger arguments one way or the other.

Reply:
I would keep it as is. I think it is best to do the check as early as possible as we ensure this way that after the genesis state initialization the NFT module state is correct and consistent.

Reply:
Will keep as is then.

Reply:
Can be resolved from my side.